core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala

```diff
@@ -187,28 +215,50 @@ private[spark] class CoarseCookSchedulerBackend(
       .setEnv(environment.asJava)
       .setUris(uris.asJava)
       .setCommand(cmds.mkString("; "))
-      .setMemory(sparkMesosScheduler.calculateTotalMemory(sc))
+      .setMemory(mesosBackend.calculateTotalMemory(sc))
       .setCpus(numCores)
       .setPriority(priority)
       .build()
   }
 
   def createRemainingJobs(): List[Job] = {
-    def loop(coresRemaining: Int, jobs: List[Job]): List[Job] =
+     @tailrec def loop(coresRemaining: Int, jobs: List[Job]): List[Job] =
       if (coresRemaining <= 0) jobs
       else if (coresRemaining <= maxCoresPerJob) createJob(coresRemaining) :: jobs
       else loop(coresRemaining - maxCoresPerJob, createJob(maxCoresPerJob) :: jobs)
-    loop(maxCores - totalCoresRequested, Nil).reverse
+
+    loop(currentCoresLimit, List()).reverse
+  }
+
+  def currentCoresLimit(): Int = {
+    jobLimitOption match {
+      case Some(num) => num * maxCoresPerJob - totalCoresRequested
+      case None => maxCores - totalCoresRequested
+    }
   }
 
   def requestRemainingCores() : Unit = {
     val jobs = createRemainingJobs()
+
     totalCoresRequested += jobs.map(_.getCpus.toInt).sum
     runningJobUUIDs = runningJobUUIDs ++ jobs.map(_.getUUID)
-    jobClient.submit(jobs.asJava, jobListener)
+
+    if (!jobs.isEmpty) jobClient.submit(jobs.asJava, jobListener)
+  }
+
+  override def doKillExecutors(executorIds: Seq[String]): Boolean = {
+    val uuids = executorIds.map(x => UUID.fromString(x)).toList
+    abortJobs(uuids)
   }
 
-  // TODO we should consider making this async, because if there are issues with cook it'll
+  override def doRequestTotalExecutors(requestedTotal: Int): Boolean = {
+    logInfo("Capping the total amount of executors to " + requestedTotal)
+    jobLimitOption = Some(requestedTotal)
+    requestRemainingCores
+    true
```

Comment:
This is the required behavior for dyn-alloc. It works the same in [mesos scheduler backend](https://github.com/mojotech/spark/blob/cook-scheduler/core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala#L402). If this method doesn't return `true` then ExecutorAllocationManager [will warn](https://github.com/mojotech/spark/blob/cook-scheduler/core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala#L358-L373) that it didn't manage to reach the cluster manager to request executors.